### DISCUSSION TITLE
Add basic set of intents for "time/tasks management" for feature parity with other voice assistants (Alexa, google home)

### DISCUSSION THREAD

**Discussion by cibernox** [home-assistant/architecture#1046] [General]
I've recently assembled an Onju Home board on a google home mini gen2 and I'm finally dabbing on voice control for my home.
In doing so, I found myself missing a lot some time management tasks that other voice assistants like Alexa (the one I'm more familiar with) or google assistant perform.

I understand that these are not strictly smart-home in the sense that they don't control any device, but nevertheless I believe that users have come to expect voice assistant to be able to perform some basic functionalities. I, for one, can tell that without some of these the WAF (wife approval factor) in my home would plummet. I can't even consider switching from Alexa to Assist without some of these:

- Asking the current time. E.g `what time is it?` -> "It's 11:22 am"
- Setting timers. E.g `Set a 20 minute timer` -> `20 minutes timer set <in device>` (and then an alarm goes off after that time. I believe user should have a sentence to dismiss the alarm too)
- Setting named timer: E.g. `Set a 12 minute timer named "boiled eggs"` -> `20 minutes timer named boiled eggs set <in device>`(and then a message/alarm goes off stating which timer is due)
- Setting alarms and named alarms: `Set an alarm for tomorrow at 9:45` / `Set an alarm for tomorrow at 9:45 named "Take Mr. Fluffy to the vet"`
- Dismissing alarms and timers before their fire.
- Inquiries about timers and alarms: Asking how much time else there is in a timer, how many timers I have left, etc...

These are the ones I miss the most, but other people might share the ones they use. I know some people set reminders and calendar events too. I also ask alexa to do math for me, specially while cooking, so I adjust the quantities of a recipe that originally was meant for 2 to make enough for 3 or more people, but let's keep this discussion around time for now.

The level of involvement of these might be very different and don't need to ship those in one go.

Some like asking what time is it are quite simple to implement, but others like adding alarms and named timers might require new services, and possibly parsing complicated time expressions like "next tuesday at 9PM'.

Are you people in favor of the general idea? If so we can make a more comprehensive list of what intents we want to support.
Are there any other basic tasks you miss from other voice assistants?

**Comment by cibernox** [comment]
Crude attemp to identify possible intents:

### `HassGetTime`
Says back the current time used the preferred time settings (24/12h, etc...)

Sentences:
- What time is it?
- What's the time right now?

### `HassSetTimer`
Setups a timer, optionally giving it a name, that will trigger a sentence or sound effect when on the same device it was set on when the time passes. That alarm continues to sound until it is dismissed, or a reasonable timeout passes.

Sentences:
- Set a 20 minutes timer
- Set a timer for 20 minutes
- Set a 20 minutes timer named muffins.

### `HassDismissTimer`
Dismisses the alarm for a timer.

Sentences:
- Cancel
- Stop the alarm
- Stop
- Shut up

### `HassGetTimers`

Sentences:
- What timers are active?
- What are the current timers?
- How many timers are there?

WIP...

**Reply by EnochPrime** [reply]
> HassGetTime
Says back the current time used the preferred time settings (24/12h, etc...)

I would also consider scenarios when triggered from the mobile app. If the intent it to create an assistant experience I would expect "what time is it?" to return the local time of where the request was made. Then "what time is it at home?" would return the Home Assistant local time.

**Reply by cibernox** [reply]
@EnochPrime that would be nice, but seems like a rather fringe situation. What I mean is that I'd do it if possible but I'd ship a simple implementation first and consider people travelling to other time zones and also asking home assistant what time is it on their phones as an improvement.

**Comment by bagobones** [comment]
Terminology for future dated items is important as there might be some differences in expected behaviors>

- set an alarm for 6pm  --> sets an unnamed alarm for that time TODAY at 6pm
- set an alarm for 6am --> sets an unnamed alarm for that time the next day at 6am
- remind me at 6pm to walk the dog --> Could add a [To-do list](https://www.home-assistant.io/integrations/todo/) item to a default Reminders list with a due time, this would also require supporting announcing these (this is how Apple Siri handles it), OR it could be a simple alarm that has a name associated to it.
- create/set an appointment for 6am tomorrow to get a hair cut ---> Could invoke the [Local Calendar](https://www.home-assistant.io/integrations/local_calendar/) and create an appointment. OR it could be a simple alarm that has a name associated to it.

in short before considering creating everything time based services a new timer of some kind we should consider if existing [To-do list](https://www.home-assistant.io/integrations/todo/)  and  [Local Calendar](https://www.home-assistant.io/integrations/local_calendar/) could be leveraged by default without expecting the end user to configure / think about it much.

**Reply by cibernox** [reply]
Yeah. I agree. First some clarification on terminology.

- Timer: It is an independent countdown with a given duration. Timers are set by specifying their duration, not the time at which they will fire, making them the easiest to implement. I do not think those need to exist in either a todo list nor a calendar. In fact I don't even think they need an UI at all. I'd 100% start with these.
- Alarms: They are set to fire at a specific time. E.g. Tomorrow at 8am. I am unconvinced that they need to be listed in a todo list nor on a calendar, but having a place in the UI to check your alarms wouldn't hurt, so I could be persuaded. This are one-time alarms, not recursive alarms meant to fire every day at a given time.

- Reminders: I can see those being akin to an entry in the calendar, but oftentime other softwares consider reminders and calendar events two different things. I wouldn't be the end of the world if we leave reminders for a future date.

**Reply by shbatm** [reply]
> Timer:... In fact I don't even think they need an UI at all.

I like the idea of the voice assistant being able to just set an existing named timer helper (or dynamically create a timer if needed--like `timer.voice_assistant_1`). I think you still need access to this in a UI and it would be best practice to use the existing infrastructures in HA. (for use case example we use Alexa for a kitchen timer and currently have a HA Dashboard that shows the timer as a flip-down card once set).

**Reply by bagobones** [reply]
HA doesn't need to exactly follow the conventions of existing assistants but I did want to call out that Google and Siri (I have both) have some behaviours that leverage existing services HA has.

I fully believe we need stand alone timer service that:
- Can create a "source specific" short lived timer IE when it ends it alerts the device that set it.
- Can create named timers, multiple timers that run from that source device. Maybe all timers are named and if not set it defaults to "timer"
- An alarm service that may just be the same service as a timer that is set not by duration but by the target end time.. Again for simplification and like the timers could be named so that more than one can be live

Some additional functionality throughs:
- Timers are probably best implemented by storing the end date/time regardless of if they were set as a duration or as an alarm at a set time.. this means the system is always just checking its stack/ queue of timers / alarms for them to become due / past due to trigger.
- Timers / alarms need to persist during restarts and probably need to FIRE if past due when HA starts but with some limitations on scope so that short HA reboots do not reset timers, and LONG offline events do not result in A LOT of timers firing all at once at startup.

**Reply by cibernox** [reply]
Yes, internally timer hey should be stored by their fire-datetime, but I wanted to highight that from a voice assistant perspective, they are set by their duration, as opposed to alarms that are set by their due date/time. 

And yes, if a HA reboots and the timer should have fired. Perhaps with some limitation tho. If you loose power during dinner having a timer for 15 minutes set and power it comes back at 3AM, maybe you don't want a speaker yelling at you for a timer that should have fired 7 hours ago.

**Reply by bagobones** [reply]
Thinking about this a any timer 1hr over due should probably just be cleaned up and logged on startup instead of fire.. Seems like a fairly safe starting point

**Comment by cibernox** [comment]
I just found out that https://github.com/home-assistant/core/pull/117199 seems to have implemented most if not all of this proposal.

@synesthesiam do you think we can close or mark this task as accepted or whatever the usual flow is?

**Reply by synesthesiam** [reply]
We're still missing a few things, like date/time and reminders. So maybe it's best to keep this open for now.

Another cool feature that should be in this release is delayed commands: "open the garage door in 5 minutes".

**Comment by aigarius** [comment]
This overlaps with https://github.com/home-assistant/architecture/discussions/1475 conceptually

The remaining part of this proposal (that heavily overlaps) would be to have a generic API to request a delayed execution of one or more actions:
* delayed action would have: name, trigger datetime, list of actions (same as automation actions, so accepting a YAML list), flag whether to do a late execution in the case that trigger time is in the past after a reboot and the actual execution datetime (for auditing already triggered actions)
* cleanup would need to delete entries that were triggered more  than long history setting ago (1 month by default)
* API to list, create, delete and edit delayed actions
* UI in Config/Tools to access the API above directly
* Automations and devices targeted by actions in pending delayed actions should have references to the action in their related section

**Reply by bagobones** [reply]
Conceptually isn't a timer for 15min or asking to turn the lights off in 15min are the same thing.. A request for a time specific event based on relative time. The only real difference is the action, play alarm, lights turn off etc.

You might actually also want to see the queue of these in the same card so you can dismiss the ones you changed your mind about.
